### PR TITLE
Add an option to force_garbage_collection for every trial.

### PR DESCRIPTION
**Pull request by oda**

### CODE REVIEW

**Comment by c-bata** [comment]
This is just a comment for core devs.

The situation that needs to call `gc.collect` is quite limited (Mark-sweep based GC may not be triggered when the environment under the memory soft limit of cgroups). The solution of #325 is enough to call `gc.collect()` at user-defined callback (refs #480). So I think it's better to just remove `gc.collect()`.

**Comment by oda** [comment]
I made this change to preserve the current behavior and to make it easier to switch the default behavior.
If you are pretty sure that nobody uses this, I am totally fine to rewrite this and just delete the gc.collect().

**Comment by c-bata** [comment]
Yes, I know why you add the option 👍  I don't oppose to merge this PR, but I think that calling `gc.collect()` is an out of the range of the responsibility of this framework. In my opinion, it should be removed or set `force_garbage_collection=False` in the future.

**Comment by g-votte** [comment]
In my opinion, it’s still worth putting `gc.collect`, because we can avoid a frequent error pattern that Optuna dev team saw multiple times in its user support experience.

Some users enbug an OOM cause and the error appears after several trials successfully ran. This kind of issue is hard to debug especially when the user is not accustomed with memory management and putting GC may be even out of his/her mind.

**Comment by c-bata** [comment]
I see. If it cannot remove `gc.collect`, this option is useful for me.

File: optuna/study.py [hunk 1]
```diff
@@ -607,6 +611,8 @@ def create_study(
             a :class:`~optuna.structs.DuplicatedStudyError` is raised if ``load_if_exists`` is
             set to :obj:`False`.
             Otherwise, the creation of the study is skipped, and the existing one is returned.
+        force_garbage_collection:
+            Flag to force gc.collect() for every trial.
```
Comment (sile) [inline optuna/study.py]: Please quote `gc.collect()` as follows:
```suggestion
            Flag to force ``gc.collect()`` for every trial.
```

File: optuna/study.py [hunk 2]
```diff
@@ -580,6 +583,7 @@ def create_study(
         study_name=None,  # type: Optional[str]
         direction='minimize',  # type: str
         load_if_exists=False,  # type: bool
+        force_garbage_collection=True,  # type: bool
```
Comment (toshihikoyanase) [inline optuna/study.py]: Please add `force_garbage_collection` to `load_study` in order to turn off GC when users resume studies.

**Pull request merged**